pytensor/tensor/rewriting/elemwise.py

```diff
@@ -529,333 +539,486 @@ def local_upcast_elemwise_constant_inputs(fgraph, node):
+    """
+    if not isinstance(node.op, Elemwise) or not isinstance(
+        node.op.scalar_op, (aes.Add, aes.Mul)
+    ):
+        return False
 
-    maker: callable
-        A function with the signature ``(node, *args)`` that constructs an
-        `op_class` instance (e.g. ``op_class(*args)``).
+    s_op = node.op.scalar_op.__class__
+    new_inp = []
+    fused = False
+    nb_inputs = len(node.inputs)
+    max_inputs = float("inf")
+    if hasattr(node.op, "max_inputs"):
+        max_inputs = node.op.max_inputs(node)
+    for inp in node.inputs:
+        if (
+            inp.owner
+            and isinstance(inp.owner.op, Elemwise)
+            and isinstance(inp.owner.op.scalar_op, s_op)
+            and
+            # Do not duplicate the operation.
+            len(fgraph.clients[inp]) == 1
+            and (nb_inputs + len(inp.owner.inputs) - 1) <= max_inputs
+        ):
+            new_inp.extend(inp.owner.inputs)
+            fused = True
+        else:
+            new_inp.append(inp)
 
-    """
-    if maker is None:
+    # We can not compare the number of inputs as Mul and Add could have
+    # 0 or 1 inputs in some corner cases.
+    if fused:
+        output = node.op(*new_inp)
+        copy_stack_trace(node.outputs[0], output)
 
-        def maker(node, scalar_op):
-            return op_class(scalar_op)
+        # Do the recursion here to help lower the number of
+        # FusionOptimizer iteration.
+        if output.owner:
+            output2 = local_add_mul_fusion.transform(fgraph, output.owner)
+            if output2:
+                return output2
+        return [output]
 
-    def local_fuse(fgraph, node):
-        r"""Fuse `Elemwise` `Op`\s in a node.
 
-        As part of specialization, we fuse two consecutive `Elemwise` `Op`\s of the
-        same shape.
+def elemwise_max_operands_fct(node) -> int:
+    # `Elemwise.perform` uses NumPy ufuncs and they are limited to 32 operands (inputs and outputs)
+    if not config.cxx:
+        return 32
+    return 1024
 
-        For mixed dtype, we let the `Composite` `Op` do the cast. It lets the C
-        compiler do the cast.
 
-        The number of dimensions is validated at call time by PyTensor itself.
+class FusionOptimizer(GraphRewriter):
+    """Graph optimizer that fuses consecutive Elemwise operations."""
 
-        """
-        # TODO: use broadcast flag?
+    def add_requirements(self, fgraph):
+        fgraph.attach_feature(ReplaceValidate())
 
-        # TODO: don't do this rewrite as a `NodeRewriter`.
-        # Analyze the graph in terms of elemwise subgraphs, and then
-        # replace each subgraph with a Composite version.
+    @staticmethod
+    def elemwise_to_scalar(inputs, outputs):
+        replace_inputs = [(inp, inp.clone()) for inp in inputs]
+        outputs = clone_replace(outputs, replace=replace_inputs)
+        # print("elemwise_to_scalar replaced outputs:")
+        # pytensor.dprint(outputs, print_type=True)
```

Comment:
forgot?

Reply:
I actually left these on purpose as I needed them a lot when debugging issues with the rewrite

Reply:
Maybe we should add a `verbosity` setting to the rewriter that can be set to turn on DEBUG level logging?

Reply:
I missed the comments in this method.